dev/com.ibm.ws.logging.core/src/com/ibm/ws/collector/manager/buffer/BufferManagerImpl.java

```diff
@@ -25,260 +26,293 @@
+            }
+
+        } finally {
+            RERWLOCK.readLock().unlock();
+        }
+
+        if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+            Tr.debug(tc, "Adding event to buffer " + event);
+        }
+
+    }
+
+    public void removeEMQ() {
+        earlyMessageQueue=null;
+    }
+
+    @Override
+    public Object getNextEvent(String handlerId) throws InterruptedException {
+        HandlerStats handlerStats = null;
+        if (handlerId != null) {
+            handlerStats = handlerEventMap.get(handlerId);
+        }
+        if (handlerStats == null)
+            throw new IllegalArgumentException("Handler not registered with buffer manager : " + handlerId);
+        // Get the next sequence number for this handler
+        long seqNum = handlerStats.getNextSeqNum();
+        Event<Object> event = ringBuffer.get(seqNum);
+        handlerStats.traceEventLoss(event.getSeqNum());
+        // Calculate the next sequence number based on the sequence number
+        // of the retrieved event.
+        long nextSeqNum = event.getSeqNum() + 1;
+        handlerStats.setNextSeqNum(nextSeqNum);
+        return event.getEvent();
+    }
+
+    @Override
+    public Object[] getEvents(String handlerId, int noOfEvents) throws InterruptedException {
+        HandlerStats handlerStats = null;
+        if (handlerId != null) {
+            handlerStats = handlerEventMap.get(handlerId);
+        }
+        if (handlerStats == null)
+            throw new IllegalArgumentException("Handler not registered with buffer manager : " + handlerId);
+        // Get the next sequence number for this handler
+        long seqNum = handlerStats.getNextSeqNum();
+        ArrayList<Event<Object>> events = ringBuffer.get(seqNum, noOfEvents);
+        handlerStats.traceEventLoss(events.get(0).getSeqNum());
+        // Calculate the next sequence number based on the sequence numbers
+        // of the retrieved events.
+        long nextSeqNum = events.get(0).getSeqNum() + events.size();
+        handlerStats.setNextSeqNum(nextSeqNum);
+        Object[] e = new Object[events.size()];
+        for (int i = 0; i < events.size(); i++) {
+            e[i] = events.get(i).getEvent();
+        }
+        return e;
+    }
+
+    public void addHandler(String handlerId) {
+        RERWLOCK.writeLock().lock();
+        try {
+            //If it is first async handler subscribed, then create the main buffer
+            if(ringBuffer == null) {
+                ringBuffer = new Buffer<Object>(capacity);
+            }
+            /*
+             * Every new Asynchronous handler starts off with all events from EMQ.
+             * So we create a copy of current EMQ and sends those messages to RingBuffer
+             */
+            try {
+                if(earlyMessageQueue != null && !earlyMessageQueue.isEmpty()) {
+                    for(Object message: earlyMessageQueue.toArray()) {
+                        ringBuffer.add(message);
```

Comment:
if ringBuffer is null you'll catch NPE, but the class will be left in weird state